aws-sdk-resources/lib/aws-sdk-resources/services/s3/encryption/decrypt_handler.rb

```diff
@@ -33,11 +38,35 @@ def attach_http_event_listeners(context)
           context.http_response.on_headers(200) do
             cipher = decryption_cipher(context)
             body = context.http_response.body
-            context.http_response.body = IODecrypter.new(cipher, body)
+            if context.http_response.headers['x-amz-meta-x-amz-tag-len']
+              # we'll need to get the entire body before decrypting
+              context.http_response.body = IODecrypter.new(cipher, body, contains_tag = true)
+            else
+              context.http_response.body = IODecrypter.new(cipher, body, contains_tag = false)
+            end
+
           end
 
           context.http_response.on_success(200) do
             decrypter = context.http_response.body
+
+            # if present, we'll need to authenticate the ciphertext
+            if context.http_response.headers['x-amz-meta-x-amz-tag-len']
+              auth_tag_len = context.http_response.headers['x-amz-meta-x-amz-tag-len'].to_i
+              decrypter.io.seek((decrypter.io.size - (auth_tag_len/8)))
+              auth_tag = decrypter.io.read
+              decrypter.cipher.auth_tag = auth_tag
+              decrypter.cipher.auth_data = ''
+              decrypter.io.truncate((decrypter.io.size - (auth_tag_len/8)))
+              decrypter.io.rewind if decrypter.io.respond_to?(:rewind)
+              # read back in the encrypted data
+              content = decrypter.io.read
+              decrypter.io.truncate(0)
+              decrypter.io.rewind if decrypter.io.respond_to?(:rewind)
+              # reprocess, decrypting data and authenticating ciphertext
+              decrypter.has_tag = false
+              decrypter.write(content)
+            end
```

Comment:
I'd like to see if we can improve the implementation to not buffer, seek, truncate and rewind the target IO object. I think this can be improved by making a 2nd GET request to the object with a range header that will retrieve only the trailing authentication tag. This can be used to initialize the cipher that we give to the IO decrypter which should eliminate the need for the extra reads.

I've pulled this down locally, and I'm going to try this out.